### PR TITLE
Adjust the index key encodings to be self-describing.

### PR DESCRIPTION
**Pull request by petermattis**

### CODE REVIEW

File: util/encoding/encoding.go [hunk 1]
```diff
@@ -493,10 +508,16 @@ func DecodeIfNull(b []byte) ([]byte, bool) {
 // empty string encoding, so do not use this routine where it is necessary to
 // distinguish not-NULL from the empty string.
 func DecodeIfNotNull(b []byte) ([]byte, bool) {
-	r := bytes.TrimPrefix(b, encodedNotNull)
-	return r, len(r) < len(b)
+	if len(b) >= 1 && b[0] == encodedNotNull {
```
Comment (tamird) [inline util/encoding/encoding.go]: i think these checks can use PeekType now

Comment (tamird) [inline util/encoding/encoding.go]: also we should file an issue to return errors from these instead of panicking

**Comment by tamird** [comment]
LGTM

File: util/encoding/encoding.go (hunk 1 above)
Comment (petermattis) [inline util/encoding/encoding.go]: Done on the `PeekType` usage. 

Plumbing through error handling everywhere these methods are used will be irritating. And I don't think I've ever seen these panics fire. I have seen other panics from the SQL code (e.g. index out of bounds). We need to make sure we have a appropriate recovery in those situations and return some sort of internal server error.

File: util/encoding/encoding.go [hunk 2]
```diff
@@ -505,17 +526,57 @@ func DecodeIfNotNull(b []byte) ([]byte, bool) {
...
+	b = append(b, timeMarker)
+	b = EncodeVarint(b, t.Unix())
+	b = EncodeVarint(b, int64(t.Nanosecond()))
 	return b
 }
 
 // DecodeTime converts an encoded time into a UTC time.Time type and
 // returns the rest of the buffer.
 func DecodeTime(b []byte) ([]byte, time.Time) {
+	if len(b) == 0 || b[0] != timeMarker {
```
Comment (tamird) [inline util/encoding/encoding.go]: also peektype

Comment (petermattis) [inline util/encoding/encoding.go]: Done.

**Pull request merged**